### PR TITLE
Core: Improve project root detection logic

### PR DESCRIPTION
**Pull request by dobesv**
- Support `.hg`
- Support `.yarn`
- Support an env var STORYBOOK_PROJECT_ROOT as an override if other detection methods are not working

Issues: #20314 #14042 #20211

## What I did

Update getProjectRoot logic to cover additional use cases.

## How to test

See the related issues for the cases that lead to this.

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

### CODE REVIEW

**Comment by kylegach** [comment]
Hey, @dobesv! Thank you for this contribution (and congratulations on your first!). Could you add some tests for this, please?

**Comment by dobesv** [comment]
Any pointers on how and where to add the tests?  I'm not too familiar with the project, and I don't think there are existing tests for this function unfortunately.  It seems a bit tricky to test.

File: code/lib/core-common/src/utils/paths.ts [hunk 1]
```diff
@@ -1,31 +1,35 @@
...
+  if (process.env.STORYBOOK_PROJECT_ROOT) {
+    return process.env.STORYBOOK_PROJECT_ROOT;
   }
-  try {
-    result = result || __dirname.split('node_modules')[0];
-  } catch (e) {
-    //
+  // Assume the innermost git/svn/hg folder is the project root
+  const foundVcsMetaDir = ['.git', '.svn', '.hg']
+    .map((dir) => findUp.sync(dir, { type: 'directory' }))
```
Comment (ndelangen) [inline code/lib/core-common/src/utils/paths.ts:12]: This is wasteful, especially considering all these are blocking the main thread sync calls.

The old code before only tried .svn if .git failed.
The new code will try all 3, even if the first one succeeds.

File: code/lib/core-common/src/utils/paths.ts [hunk 2]
```diff
@@ -1,31 +1,35 @@
...
+  for (;;) {
+    const cwd = projectRoot;
+    const foundNodeModulesDir = ['node_modules', '.yarn']
+      .map((dir) => findUp.sync(dir, { cwd, type: 'directory' }))
+      .filter(Boolean)[0];
+    if (foundNodeModulesDir) {
+      projectRoot = path.join(foundNodeModulesDir, '..');
+    } else {
+      break;
+    }
```
Comment (ndelangen) [inline code/lib/core-common/src/utils/paths.ts:30]: Won't a split on a regex be able to essentially do the same thing, without using findUp?

**Review by ndelangen — CHANGES_REQUESTED** [review]
I left 2 comments

File: code/lib/core-common/src/utils/paths.ts (hunk 2 above)
Comment (dobesv) [reply code/lib/core-common/src/utils/paths.ts:30]: Hmm good point.

File: code/lib/core-common/src/utils/paths.ts (hunk 1 above)
Comment (dobesv) [reply code/lib/core-common/src/utils/paths.ts:12]: True

**Comment by ndelangen** [comment]
@dobesv thank you for contributiong! I think this change is good, but I'd like it implemented in a slightly different way, I hope you can make that happen? 🙏

**Comment by dobesv** [comment]
Thanks for the review @ndelangen .  I made changes based on your feedback.  If you have any thoughts about how I could make automated tests for this without a ton of work, let me know.

**Review by ndelangen — APPROVED** [review]

**Comment by ndelangen** [comment]
We could add tests that mock `find-up`; that would mean we only have to specify when find-up with return what string?

**Review by ndelangen — APPROVED** [review]

**Pull request merged**